scripts/scr_get_item_names/scr_get_item_names.gml

```diff
@@ -0,0 +1,806 @@
+/// @param {bool} _melee_or_ranged - Whether the equipment is melee or ranged. true for melee, false for ranged.
+/// @param {bool} _is_master_crafted - Whether to include only master-crafted items.
+/// @param {array} _required_tags - Tags that the equipment must have.
+/// @param {array} _excluded_tags - Tags that the equipment must not have.
+/// @param {bool} _with_none - Include "(None)" in the list.
+/// @param {bool} _with_any - Include "(any)" in the list.
+/// @returns {array} item_names - The filtered list of equipment names.
+function get_filtered_equipment_item_names(_equip_category, _melee_or_ranged, _is_master_crafted=false, _required_tags=undefined, _excluded_tags=undefined, _with_none=false, _with_any=false) {
+    var _item_names = get_none_or_any_item_names(_with_none, _with_any);
+    for (var _i = 0; _i < array_length(obj_ini.equipment); _i++) {
+        if (_is_master_crafted && !array_contains(obj_ini.equipment_quality[_i], "master_crafted")) {
+            continue;
+        }
+        var equip_data = gear_weapon_data(_equip_category, obj_ini.equipment[_i]);
+        if (is_struct(equip_data) && obj_ini.equipment_number[_i] > 0) {
+            if (_melee_or_ranged != undefined) {
+                if (_melee_or_ranged && equip_data.range > 1.1) {
+                    continue;
+                } else if (!_melee_or_ranged && equip_data.range <= 1.1) {
+                    continue;
+                }
+            }
+			var valid = true;
+            if (_required_tags != undefined) {
+                for (var _t = 0; _t < array_length(_required_tags); _t++) {
+                    if (!equip_data.has_tag(_required_tags[_t])) {
+                        valid = false;
+                        break;
+                    }
+                }
+            }
+            if (_excluded_tags != undefined) {
+                for (var _t = 0; _t < array_length(_excluded_tags); _t++) {
+                    if (equip_data.has_tag(_excluded_tags[_t])) {
+                        valid = false;
+                        break;
+                    }
+                }
+            }
+			if (valid) {
+				array_push(_item_names, equip_data.name);
+			}
+        }
+    }
+    return _item_names;
+}
+
+enum eEQUIPMENT_TYPE {
+    None,
+    PrimaryWeapon = 1,  // LeftHand, Turret, Front, Primary
+    SecondaryWeapon = 2,  // RightHand, Sponson, Secondary
+    Armour = 3,
+    GearUpgrade = 4,
+    MobilityAccessory = 5
+}
+
+enum eEQUIPMENT_SUBTYPE {
+    None,
+    Ranged = 1, // Regular land raider weapons
+    Melee = 2 // Relic land raider weapons
+}
+
+enum eUNIT_TYPE {
+    None,
+    Infantry = 1,
+    Dreadnought = 6,
+    LandRaider = 50,
+    Rhino = 51,
+    Predator = 52,
+    LandSpeeder = 53,
+    Whirlwind = 54
+}
+
+// slot names differ by unit type then equipment type
+// HACK: GM doesn't like strings that look like numbers as struct keys
+//  so we prefix them with an underscore
+global.slot_names = {
```

Comment:
If this variable is not intended to change at run-time, is it impossible to use an enum?
Or, considering that atm it's used only in one function, contain it locally in that function?

I just try to avoid global variables as much as possible, as I've seen notes that they are very bad practice from a performance standpoint. In this case it may be okay, but if it's possible to avoid, then better to avoid.

Reply:
These are not intended to change at run time.

I'm not sure what structure you should use in gml to have a constant lookup of strings based on two enums. Applying my python brain, I'd make a global dictionary with two levels and call it a day. In javascript, I'd make it a local in the outer closure of the function and access it inside.

Should I convert it to a switch inside the lookup function and delete the global?

Reply:
>  Applying my python brain, I'd make a global dictionary with two levels and call it a day. In javascript, I'd make it a local in the outer closure of the function and access it inside.

I'm not sure if I'm misremembering, but I've seen people saying that you should only use `globals` when there is no way to get away without them and not turn your code unreadable/inefficient.

> Should I convert it to a switch inside the lookup function and delete the global?

I'll suggest a couple of options, that I think may be viable, then you can decide on your own. I just can't explain it with words, as I need to fiddle with code.

Reply:
Well, yeah. I came to this, which is your suggestion with the switch.
I'm not sure if "eEQUIPMENT_TYPE" enum is nessessary, because slots are just slots. First, second, etc.
But this is just how I think.
```suggestion
/// @description This function returns the name of the slot for a given unit type and equipment type.
/// @param {real} _unit_type - The type of unit to equip, see eUNIT_TYPE.
/// @param {real} _equipment_slot - The equipment slot number.
/// @returns {string} The name of the slot.
function get_slot_name(_unit_type, _equipment_slot) {
    // Map unit types to slot name maps
    switch (_unit_type) {
        case eUNIT_TYPE.Infantry:
            switch (_equipment_slot) {
                case 1: return "Left Hand";
                case 2: return "Right Hand";
                case 3: return "Armour";
                case 4: return "Gear";
                case 5: return "Mobility";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Dreadnought:
            switch (_equipment_slot) {
                case 1: return "Left Arm";
                case 2: return "Right Arm";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.LandRaider:
            switch (_equipment_slot) {
                case 1: return "Front";
                case 2: return "Sponson";
                case 3: return "Pintle";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Rhino:
            switch (_equipment_slot) {
                case 1: return "Weapon";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Predator:
            switch (_equipment_slot) {
                case 1: return "Turret";
                case 2: return "Sponsons";
                case 3: return "Pintle";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.LandSpeeder:
            switch (_equipment_slot) {
                case 1: return "Primary";
                case 2: return "Secondary";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Whirlwind:
            switch (_equipment_slot) {
                case 1: return "Missiles";
                case 2: return "Pintle";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        default:
            return "Unknown";
    }
    return "Unknown";
}
```

Reply:
There are probably some interesting ways of doing this with ds_ stuff, but I have 0 experience with them and just ignore this part of GM, because they are never garbage collected and have to be perfectly coded.

Reply:
I switched to using a switch in `get_slot_name` in ee82b82d.